internal/rbd/controllerserver.go

```diff
@@ -332,6 +310,47 @@ func flattenParentImage(ctx context.Context, rbdVol *rbdVolume, cr *util.Credent
 	return nil
 }
 
+// repairExistingVolume checks the existing volume or snapshot and makes sure
+// that the state is corrected to what was requested. It is needed to call this
+// when the process of creating a volume was interrupted.
+func (cs *ControllerServer) repairExistingVolume(ctx context.Context, req *csi.CreateVolumeRequest,
+	cr *util.Credentials, rbdVol *rbdVolume, rbdSnap *rbdSnapshot) (*csi.CreateVolumeResponse, error) {
+	vcs := req.GetVolumeContentSource()
+
+	switch {
+	// normal CreateVolume without VolumeContentSource
+	case vcs == nil:
+		// continue/restart allocating the volume in case it
+		// should be thick-provisioned
+		if cs.isThickProvisionRequest(req) {
```

Comment:
Just a corner case. If the allocation is completed but failed to set metadata due to restart. we are going to do the allocation again. can we check the used size of the image is equal and if yes just set the metadata or if no continue the allocation?

Reply:
Yes, there is a `TODO` noted in the `RepairThickProvision()` function. It would be an optimization for the case where a recurring error happened, so I do not deem this important to address.